assets/js/utils/rtcUtils.ts

```diff
@@ -1,57 +1,140 @@
 import { MembraneWebRTC, SerializedMediaEvent } from "@membraneframework/membrane-webrtc-js";
-import { presenterStreams } from "../components/PresenterStreamArea";
 
-export const AUDIO_CONSTRAINTS: MediaStreamConstraints = {
+export type Sources = {
+  audio: MediaDeviceInfo[];
+  video: MediaDeviceInfo[];
+};
+
+export type SourceType = "audio" | "video";
+
+export const presenterStreams: { [key: string]: MediaStream } = {};
+
+export const findTrackByType = (name: string, sourceType: SourceType) => {
+  return presenterStreams[name].getTracks().find((elem) => elem.kind == sourceType);
+};
+
+const addOrReplaceTrack = (
+  name: string,
+  track: MediaStreamTrack,
+  playerCallback: (sourceType: SourceType) => void
+) => {
+  if (!presenterStreams[name]) presenterStreams[name] = new MediaStream();
+  const curTrack = findTrackByType(name, track.kind as SourceType);
+  if (curTrack) {
+    curTrack.stop();
+    presenterStreams[name].removeTrack(curTrack);
+  }
+  presenterStreams[name].addTrack(track);
+  playerCallback(track.kind as SourceType); // to attach MediaStream to HTMLVideoElement object in DOM
+};
+
+const removeStream = (name: string) => {
+  delete presenterStreams[name];
+};
+
+export const changeTrackIsEnabled = (name: string, sourceType: SourceType) => {
+  const track = findTrackByType(name, sourceType);
+  if (track) track.enabled = !track.enabled;
+};
+
+export const getCurrentDeviceName = (clientName: string, sourceType: SourceType) => {
+  return findTrackByType(clientName, sourceType)?.label;
+};
+
+const AUDIO_CONSTRAINTS: MediaStreamConstraints = {
   audio: true,
   video: false,
 };
 
-export const VIDEO_CONSTRAINTS: MediaStreamConstraints = {
+const VIDEO_CONSTRAINTS: MediaStreamConstraints = {
   audio: false,
   video: { width: 1280, height: 720, frameRate: 24 },
 };
 
-export const connectWebrtc = async (
-  webrtcChannel: any,
-  name: string,
-  streamsAvailable: { [key: string]: boolean },
-  setStreamsAvailable: React.Dispatch<React.SetStateAction<{ [key: string]: boolean }>>
-) => {
-  let localAudioStream: MediaStream | null = null;
-  let localVideoStream: MediaStream | null = null;
-  let localStream: MediaStream = new MediaStream();
+const getConstraint = (constraint: MediaStreamConstraints, deviceId: string) => {
+  const newConstraint: MediaStreamConstraints = { audio: {}, video: {} };
+  let type: SourceType;
+  if (!constraint.audio) {
+    newConstraint.audio = false;
+    type = "video";
+  } else {
+    newConstraint.video = false;
+    type = "audio";
+  }
+  const elem = constraint[type] === true ? {} : constraint[type];
+  elem!["deviceId"] = { exact: deviceId };
+  newConstraint[type] = elem;
 
+  return newConstraint;
```

Comment:
I'm trying to avoid TypeScript warning that only object types can be spread (constraint[type] is of MediaStreamConstraint type). Also the non-false constraint can be a boolean which line 64 prevents. Still refactored slightly.

Reply:
If there are TypeScript warnings we could cast this variable to an object. But if you think this will be too big a hack, I advise doing some more refactoring because it is a little bit confusing still for me.